// @supercolony/typechain-polkadot-parser/src/utils.ts
// Copyright (c) 2012-2022 Supercolony
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the"Software"),
// to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import {INK_TYPES_TO_TS_ARGUMENTS, INK_TYPES_TO_TS_RETURNS} from "./consts";
import assert from "assert";
import {Abi} from "@polkadot/api-contract";
import {TypeInfo, TypeTS} from "./types/TypeInfo";
import camelcase from "camelcase";

export const parsePrimitiveReturns = (primitive: string): string => {
	// @ts-ignore
	return INK_TYPES_TO_TS_RETURNS[primitive];
};

export const parsePrimitiveArgs = (primitive: string): string => {
	// @ts-ignore
	return INK_TYPES_TO_TS_ARGUMENTS[primitive];
};

export const generateInterfaceReturns = (interfaceName: string, argumentNames: string[], argumentTypes: TypeInfo[]) => {
	assert(argumentNames.length == argumentTypes.length);

	if (argumentNames.filter(a => a === '').length > 0) {
		if (argumentNames.length === 1) {
			return `export type ${interfaceName} = ${argumentTypes[0]!.tsReturnType};`;
		} else {
			return `export type ${interfaceName} = [${argumentTypes.map(a => a!.tsReturnType).join(', ')}];`;
		}
	}

	return `export type ${interfaceName} = {
\t${argumentNames.map((e, i) => `${camelcase(e)}: ${argumentTypes[i]!.tsReturnType}`).join(',\n\t')}
}`;
};

export const generateInterfaceArgs = (interfaceName: string, argumentNames: string[], argumentTypes: TypeInfo[]) => {
	assert(argumentNames.length == argumentTypes.length);

	if (argumentNames.filter(a => a === '').length > 0) {
		if (argumentNames.length === 1) {
			return `export type ${interfaceName} = ${argumentTypes[0]!.tsArgType};`;
		} else {
			return `export type ${interfaceName} = [${argumentTypes.map(a => a!.tsArgType).join(', ')}];`;
		}
	}

	return `export type ${interfaceName} = {
\t${argumentNames.map((e, i) => `${camelcase(e)}: ${argumentTypes[i]!.tsArgType}`).join(',\n\t')}
}`;
};

export const generateEnum = (enumName: string, enumFields: string[]): string => {
	return `export enum ${enumName} {
	${enumFields.map(e => `${camelcase(e)} = '${e}'`).join(',\n\t')}
}`;
};

export const generateClassEnum = (enumName: string, enumFields: string[], enumValues: string[]): string => {
	assert(enumFields.length == enumValues.length);
	return `export interface ${enumName} {
	${enumFields.map((e, i) => `${camelcase(e)} ? : ${enumValues[i]}`).join(',\n\t')}
}

export class ${enumName}Builder {
	${enumFields.map((e, i) => `static ${e}(${enumValues[i] !== 'null' ? `value: ${enumValues[i]}` : ''}): ${enumName} {
		return {
		${enumValues[i] !== 'null' ? `\t${camelcase(e)}: value` : `\t${camelcase(e)}: null`},
		};
	}`).join('\n\t')}
}`;
};

/**
 * Preprocesses the ABI to avoid changes from polkadot.js
 *
 * @param _abiStr - The ABI to preprocess
 * @returns The preprocessed ABI
 */
export function preprocessABI(_abiStr: string): Abi {
	const abiJson = JSON.parse(_abiStr);

	for (const method of abiJson.V3.spec.messages) {
		for (const arg of method.args) {
			for (let i = 0; i < arg.type.displayName.length; i++) {
				arg.type.displayName[i] = `_${arg.type.displayName[i]}`;
			}
		}
	}

	const typeNamesCount = new Map<string, number>();

	for (const {type} of abiJson.V3.types) {
		if (type.path === undefined) continue;
		if (type.path[type.path.length - 1] == 'Mapping') continue;

		if (type.path.length > 0) {
			const value = typeNamesCount.get(type.path[type.path.length - 1]) || 0;
			typeNamesCount.set(
				type.path[type.path.length - 1],
				value + 1
			);
		}
	}

	let __i = 0;
	for (const {type} of abiJson.V3.types) {
		__i++;
		if (type.path === undefined) continue;
		if (type.path[type.path.length - 1] == 'Mapping') continue;

		const count = typeNamesCount.get(type.path[type.path.length - 1]);
		if (type.path.length > 0 && (count ? count : 0) > 1) {
			if (type.path.length > 3) {
				abiJson.V3.types[__i - 1].type.path[type.path.length - 1] = `${type.path[type.path.length - 2]}_${type.path[type.path.length - 1]}`;
			}
		}
	}

	const _abiStrWithUnderscores = JSON.stringify(abiJson, null, 2);

	return new Abi(_abiStrWithUnderscores);
}
