// @supercolony/typechain-types/dist/src/tx.js
"use strict";
// Copyright (c) 2012-2022 Supercolony
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the"Software"),
// to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._signAndSend = exports.buildSubmittableExtrinsic = exports.txSignAndSend = void 0;
var query_1 = require("./query");
function txSignAndSend(nativeAPI, nativeContract, keyringPair, title, eventHandler, args, gasLimitAndValue) {
    var submittableExtrinsic = buildSubmittableExtrinsic(nativeContract, title, args, gasLimitAndValue);
    return _signAndSend(nativeAPI.registry, submittableExtrinsic, keyringPair, eventHandler);
}
exports.txSignAndSend = txSignAndSend;
function buildSubmittableExtrinsic(nativeContract, title, args, gasLimitAndValue) {
    var _a;
    if (nativeContract.tx[title] == null) {
        var error = {
            issue: 'METHOD_DOESNT_EXIST',
            texts: ["Method name: '".concat(title, "'")],
        };
        throw error;
    }
    var _args = args || [];
    var _gasLimitAndValue = (0, query_1._genValidGasLimitAndValue)(gasLimitAndValue);
    var submittableExtrinsic = (_a = nativeContract.tx)[title].apply(_a, __spreadArray([_gasLimitAndValue], _args, false));
    return submittableExtrinsic;
}
exports.buildSubmittableExtrinsic = buildSubmittableExtrinsic;
/**
 * (i) For reference, see:
 * 	- https://polkadot.js.org/docs/api/cookbook/tx#how-do-i-get-the-decoded-enum-for-an-extrinsicfailed-event
 * 	- `@redspot/patract/buildTx`
 */
function _signAndSend(registry, extrinsic, signer, eventHandler) {
    return __awaiter(this, void 0, void 0, function () {
        var signerAddress;
        return __generator(this, function (_a) {
            signerAddress = signer.address;
            return [2 /*return*/, new Promise(function (resolve, reject) {
                    var actionStatus = {
                        from: signerAddress.toString(),
                        txHash: extrinsic.hash.toHex(),
                    };
                    extrinsic
                        .signAndSend(signer, function (result) {
                        if (result.status.isInBlock) {
                            actionStatus.blockHash = result.status.asInBlock.toHex();
                        }
                        if (result.status.isFinalized || result.status.isInBlock) {
                            actionStatus.events = eventHandler(result.events);
                            result.events
                                .filter(function (_a) {
                                var section = _a.event.section;
                                return section === 'system';
                            })
                                .forEach(function (event) {
                                var _a = event.event, data = _a.data, method = _a.method;
                                if (method === 'ExtrinsicFailed') {
                                    var dispatchError = data[0];
                                    var message = dispatchError.type;
                                    if (dispatchError.isModule) {
                                        try {
                                            var mod = dispatchError.asModule;
                                            var error = registry.findMetaError(new Uint8Array([
                                                mod.index.toNumber(),
                                                mod.error.toNumber()
                                            ]));
                                            message = "".concat(error.section, ".").concat(error.name).concat(Array.isArray(error.docs)
                                                ? "(".concat(error.docs.join(''), ")")
                                                : error.docs || '');
                                        }
                                        catch (error) {
                                            // swallow
                                        }
                                    }
                                    actionStatus.error = {
                                        message: message,
                                    };
                                    reject(actionStatus);
                                }
                                else if (method === 'ExtrinsicSuccess') {
                                    actionStatus.result = result;
                                    resolve(actionStatus);
                                }
                            });
                        }
                        else if (result.isError) {
                            actionStatus.error = {
                                data: result,
                            };
                            actionStatus.events = undefined;
                            reject(actionStatus);
                        }
                    })
                        .catch(function (error) {
                        actionStatus.error = {
                            message: error.message,
                        };
                        reject(actionStatus);
                    });
                })];
        });
    });
}
exports._signAndSend = _signAndSend;
